doc/source/reference/c-api/array.rst

```diff
@@ -1555,8 +1555,296 @@ For all of these macros *arr* must be an instance of a (subclass of)
+documentation on the C structs exposed in the ArrayMethod API.
+
+Slots
+~~~~~
+
+These are used to identify which kind of function an ArrayMethod slot
+implements. See :ref:`arraymethod-typedefs` below for documentation on
+the functions that must be implemented for each slot.
+
+.. c:macro:: NPY_METH_resolve_descriptors
+
+   The function used to set the descriptors for an operation based on the
+   descriptors of the operands. For example, a ufunc operation with two input
+   operands and one output operand that is called without ``out`` being set
+   in the python API, ``resolve_descriptors`` will be passed the descriptors
+   for the two operands and determine the correct descriptor to use for the
+   output based on the output DType set for the ArrayMethod. If ``out`` is
+   set, then the output descriptor would be passed in as well and should not
+   be overriden.
+
+.. c:macro:: NPY_METH_get_reduction_initial
+
+   A function that gets the initial value to use for a reduction.
+
+.. c:macro:: NPY_METH_strided_loop
+.. c:macro:: NPY_METH_contiguous_loop
+.. c:macro:: NPY_METH_unaligned_strided_loop
+.. c:macro:: NPY_METH_unaligned_contiguous_loop
+
+   One dimensional strided loops implementing the behavior (either a ufunc
+   or cast).  In most cases, ``NPY_METH_strided_loop`` is the generic and only
+   version that needs to be implemented.
+   ``NPY_METH_contiguous_loop`` can be implemented additionally as a more
+   light-weight/faster version and it is used when all inputs and outputs
+   are contiguous.
+
+   To deal with possibly unaligned data, NumPy needs to be able to copy
+   unaligned to aligned data.  When implementing a new DType, the "cast" or
+   copy for it needs to implement ``NPY_METH_unaligned_strided_loop``.
+   Unlike the normal versions, this loop must not assume that the data can be
+   accessed in an aligned fashion.  These loops must copy each value before
+   accessing or storing::
+
+       type_in in_value;
+       type_out out_value
+       memcpy(&value, in_data, sizeof(type_in));
+       out_value = in_value;
+       memcpy(out_data, &out_value, sizeof(type_out)
+
+   while a normal loop can just use::
+
+       *(type_out *)out_data = *(type_in)in_data;
+
+   The unaligned loops are currently only used in casts and will never be picked
+   in ufuncs (ufuncs create a temporary copy to ensure aligned inputs).
+   These flags are ignored when ``NPY_METH_get_loop`` is defined, where instead
+   whichever loop returned by the ``get_loop`` function is used.
+
+.. c:macro:: NPY_METH_contiguous_indexed_loop
+
+   A specialized inner-loop option to speed up common ``ufunc.at`` computations.
+
+.. c:macro:: NPY_METH_get_loop
+
+   Allows more fine-grained control over loop selection. Accepts an
+   implementation of PyArrayMethod_GetLoop, which in turn returns a
+   strided loop implementation. If ``NPY_METH_get_loop`` is defined,
+   the other loop slot IDs are ignored, if specified.
```

Comment:
This implies also the indexed loop is ignored, correct?

Reply:
Unfortunately not.  TBH, I wouldn't be unhappy to just not expose the indexed loop for now.

Reply:
Either way is fine, but the text should be correct!

Reply:
reworded